File: src/295.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class MedianFinder:
    def __init__(self):
        self.minHeap = []
        self.maxHeap = []

    def addNum(self, num: int) -> None:
        heappush(self.maxHeap, -num)
        heappush(self.minHeap, -heappop(self.maxHeap))

        maxHeapSize = len(self.maxHeap)
        minHeapSize = len(self.minHeap)
        if minHeapSize > maxHeapSize:
            heappush(self.maxHeap, -heappop(self.minHeap))

    def findMedian(self) -> float:
        retVal = -self.maxHeap[0]

        maxHeapSize = len(self.maxHeap)
        minHeapSize = len(self.minHeap)
        if maxHeapSize > minHeapSize:
            return retVal
        retVal = (-self.maxHeap[0] + self.minHeap[0]) / 2

        return retVal


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        logging.info("Input")
        for methods, num in zip([["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"]],
                                [[None, 1, 2, None, 3, None]]):
            # /* Example
            #  *  Input
            #  *  ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"]
            #  *  [[], [1], [2], [], [3], []]
            #  *  Output
            #  *  [null, null, null, 1.5, null, 2.0]
            #  */
            logging.info("%s", methods)
            displayStr = []
            for p in num:
                if p == None:
                    displayStr.append([])
                else:
                    displayStr.append(p)
            logging.info("%s", displayStr)

            displayStr = []
            obj = MedianFinder()
            for p, q in zip(methods, num):
                if p == "MedianFinder":
                    displayStr.append(None)
                elif p == "addNum":
                    obj.addNum(q)
                    displayStr.append(None)
                elif p == "findMedian":
                    retVal = obj.findMedian()
                    displayStr.append(retVal)
            logging.info("Output")
            logging.info("%s", displayStr)

            print("")
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/295.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class MedianFinder {
   public:
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;

    MedianFinder() {}
    void addNum(int num) {
        maxHeap.push(num);
        minHeap.push(maxHeap.top());
        maxHeap.pop();

        int maxHeapSize = maxHeap.size();
        int minHeapSize = minHeap.size();
        if (minHeapSize > maxHeapSize) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }
    double findMedian() {
        double retVal = maxHeap.top();

        int maxHeapSize = maxHeap.size();
        int minHeapSize = minHeap.size();
        if (maxHeapSize > minHeapSize) {
            return retVal;
        }
        retVal = (maxHeap.top() + minHeap.top()) / 2.0;

        return retVal;
    }
};
/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

int main(int argc, char** argv) {
    struct subject {
        vector<string> methods;
        vector<int> num;
    };
    vector<subject> testData{
        {{"MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"}, {0, 1, 2, 0, 3, 0}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input
     *  ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"]
     *  [[], [1], [2], [], [3], []]
     *  Output
     *  [null, null, null, 1.5, null, 2.0]
     */

    for (int i = 0; i < numberOfTestCase; ++i) {
        MedianFinder cMedianFinder;

        cout << "Input\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].methods.size(); ++j) {
            cout << (j == 0 ? "" : ", ") << '"' << testData[i].methods[j] << '"';
        }
        cout << "]\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].num.size(); ++j) {
            cout << (j == 0 ? "" : ", ") << "[";
            if (testData[i].num[j] != 0) {
                cout << testData[i].num[j];
            }
            cout << "]";
        }
        cout << "]\n";

        cout << "Output\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].methods.size(); ++j) {
            if (testData[i].methods[j].compare("MedianFinder") == 0) {
                cout << (j == 0 ? "" : ", ") << "null";
            } else if (testData[i].methods[j].compare("addNum") == 0) {
                cMedianFinder.addNum(testData[i].num[j]);
                cout << (j == 0 ? "" : ", ") << "null";
            } else if (testData[i].methods[j].compare("findMedian") == 0) {
                double retVal = cMedianFinder.findMedian();
                cout.setf(ios::fixed);
                cout.precision(1);
                cout << (j == 0 ? "" : ", ") << retVal;
            }
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
